conjure-http/src/client.rs

```diff
@@ -223,3 +237,219 @@ pub trait AsyncWriteBody<W> {
     where
         W: 'async_trait;
 }
+
+/// A trait implemented by request body serializers used by custom Conjure client trait
+/// implementations.
+pub trait SerializeRequest<'a, T, W> {
+    /// Returns the body's content type.
+    fn content_type(value: &T) -> HeaderValue;
+
+    /// Returns the body's length, if known.
+    ///
+    /// Empty and fixed size bodies will have their content length filled in automatically.
+    ///
+    /// The default implementation returns `None`.
+    fn content_length(value: &T) -> Option<u64> {
+        let _value = value;
+        None
+    }
+
+    /// Serializes the body.
+    fn serialize(value: T) -> Result<RequestBody<'a, W>, Error>;
+}
+
+/// A trait implemented by request body serializers used by custom async Conjure client trait
+/// implementations.
+pub trait AsyncSerializeRequest<'a, T, W> {
+    /// Returns the body's content type.
+    fn content_type(value: &T) -> HeaderValue;
+
+    /// Returns the body's length, if known.
+    ///
+    /// Empty and fixed size bodies will have their content length filled in automatically.
+    ///
+    /// The default implementation returns `None`.
+    fn content_length(value: &T) -> Option<u64> {
+        let _value = value;
+        None
+    }
+
+    /// Serializes the body.
+    fn serialize(value: T) -> Result<AsyncRequestBody<'a, W>, Error>;
+}
+
+/// A body serializer which encodes the type into JSON.
+///
+/// # Note
+///
+/// The default `serde_json` serializer is used, *not* the `conjure-serde` serializer.
+pub enum JsonRequestSerializer {}
```

Comment:
Maybe a dumb question, but why is this an enum vs a struct?

Reply:
An empty enum can't be constructed, which we don't need to do in this context since none of the methods on the trait take `self`.